### PR TITLE
Suppress zero-length text SVG inline renderers

### PR DESCRIPTION
**Pull request by Ahmad-S792**
<pre>
Suppress zero-length text SVG inline renderers
<a href="https://bugs.webkit.org/show_bug.cgi?id=119719">https://bugs.webkit.org/show_bug.cgi?id=119719</a>

Reviewed by NOBODY (OOPS!).

Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/e4c158a33d34264796abfa683d0413568dfdcde0">https://chromium.googlesource.com/chromium/blink/+/e4c158a33d34264796abfa683d0413568dfdcde0</a>

The SVG text layout functions assert non-zero metrics for SVG inline text (see SVGTextLayoutEngine::layoutTextOnLineOrPath, for example). 
Normally, we don't instantiate renderers for zero-length inline text (see Text::textRendererIsNeeded), but the RenderSVGInlineText constructor 
performs whitespace filtering to support xml:space semantics (<a href="http://www.w3.org/TR/SVG/struct.html#LangSpaceAttrs">http://www.w3.org/TR/SVG/struct.html#LangSpaceAttrs</a>) 
and can end up with empty text while still passing the non-zero length test in Text::textRendererIsNeeded 
(for example a single newline gets deleted when xml:space="default").

Enter RenderText::positionLineBox() which attempts to fix things up by removing zero length
inline boxes. But removing boxes at this point is bad for business because it invalidates
(marks as dirty and clears computed metrics) parent boxes and sibling boxes during layout,
and throws subsequent computations off.

The patch adds isChildAllowed()-based validation for SVG inline text. This prevents 
zero-len-text renderers from being inserted into the tree in the first place.

Also removing a redundant test in SVGTextLayoutEngine::currentVisualCharacterMetrics():
the while loop condition handles that case in an equivalent manner.

* Source/WebCore/rendering/RenderText.h: Add 'hasEmptyText' bool function
* Source/WebCore/rendering/svg/RenderSVGInline.cpp:
(RenderSVGInline::isChildAllowed): New function
* Source/WebCore/rendering/svg/RenderSVGInline.h: Add Function definition 'isChildAllowed'
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(RenderSVGText::isChildAllowed): Update return
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp: Add 'isEmptySVGInlineText' bool function
* Source/WebCore/rendering/svg/SVGRenderSupport.h: Add Function definition 'isEmptySVGInlineText'
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(SVGTextLayoutEngine::currentVisualCharacterMetrics): Remove redundant case
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-altglyph-01-b-manual-expected.txt: Rebaselined
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-altglyph-01-b-manual-expected.txt: Rebaselined
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/import/text-altglyph-01-b-manual-expected.txt: Rebaselined
* LayoutTests/svg/custom/invalid-text-content-expected.txt: Rebaselined
* LayoutTests/svg/text/font-size-below-point-five-expected.txt: Rebaselined
* LayoutTests/svg/text/textPathBoundsBug-expected.txt: Rebaselined
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/text-altglyph-01-b-expected.txt: Rebaselined
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/text-align-08-b-expected.txt: Rebaselined
* LayoutTests/platform/ios/svg/text/text-hkern-expected.txt: Rebaselined
* LayoutTests/platform/ios/svg/text/text-altglyph-01-b-expected.txt: Rebaselined
* LayoutTests/platform/ios/svg/text/font-size-below-point-five-expected.txt: Rebaselined
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-align-08-b-expected.txt: Rebaselined
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-align-08-b-expected.png: Rebaselined
* LayoutTests/platform/glib/svg/text/font-size-below-point-five-expected.txt: Rebaselined
* LayoutTests/platform/glib/svg/text/textPathBoundsBug-expected.txt: Rebaselined
* LayoutTests/platform/gtk/svg/text/font-size-below-point-five-expected.png: Rebaselined
* LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png: Rebaselined
* LayoutTests/platform/mac/svg/text/font-size-below-point-five-expected.png: Rebaselined
* LayoutTests/platform/mac/svg/text/font-size-below-point-five-expected.txt: Rebaselined
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.png: Rebaselined
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.txt: Rebaselined
* LayoutTests/platform/wpe/svg/text/text-altglyph-01-b-expected.txt: Rebaselined
* LayoutTests/platform/wpe/svg/text/text-hkern-expected.txt: Rebaselined
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-align-08-b-expected.txt: Rebaselined
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-altglyph-01-b-expected.txt: Rebaselined
</pre>
https://github.com/WebKit/WebKit/commit/110c9fa86e859cf29a5a12fb45159d3b2b68448d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21531 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22359 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19091 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20692 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20491 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20549 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17789 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23211 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18594 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24880 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18770 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22820 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19360 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16442 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18573 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22913 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19181 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png
Comment (nikolaszimmermann) [inline LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png:1]: I've spotted a visual change here, seems like the grey-ish selection rect is missing in the new result? Is that a regression of yours?

File: LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.png
Comment (nikolaszimmermann) [inline LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.png:1]: What is changing here, just a few px diff? Can't spot it...

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp [hunk 1]
```diff
@@ -360,9 +360,6 @@ bool SVGTextLayoutEngine::currentVisualCharacterMetrics(const SVGInlineTextBox&
     unsigned boxStart = textBox.start();
     unsigned boxLength = textBox.len();
 
-    if (m_visualMetricsListOffset == textMetricsSize)
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:363]: Hmm, not fully understand the impact of this change. Can you elaborate?

File: Source/WebCore/rendering/svg/SVGRenderSupport.h [hunk 2]
```diff
@@ -46,6 +47,8 @@ class SVGRenderSupport {
 public:
     static void layoutDifferentRootIfNeeded(const RenderElement&);
 
+    static bool isEmptySVGInlineText(const RenderObject*);
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGRenderSupport.h:50]: I know you've cherry-picked a Blink change, but I suggest to not add anything new, valuable to SVGRenderSupport, which is a messy legacy container class. This should go in another place.
I'd vote for using dynamicDowncast<> on the call side, and just avoid the helper alltogether.

**Review by nikolaszimmermann — COMMENTED** [review]
Thanks @Ahmad-S792! Almost r+, one more iteration please :-)

File: LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.png
Comment (Ahmad-S792) [reply LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-align-08-b-expected.png:1]: Yes - it is just 0.04% pixel difference and not visible.

File: LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png
Comment (Ahmad-S792) [reply LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png:1]: Will update test to use 'Selection API' similar to Chromium / Blink. Which will fix this issue. :-)

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:363]: This is about following ```removing a redundant test in SVGTextLayoutEngine::currentVisualCharacterMetrics():
the while loop condition handles that case in an equivalent manner.``` from the commit, 'while' loop just below takes care of it.

File: Source/WebCore/rendering/svg/SVGRenderSupport.h (hunk 2 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGRenderSupport.h:50]: I looked into Blink code and they have something like:

Or I am not clear on 'another place', any idea, where I could move this?

Blink latest code look something like this:

```
bool SVGRenderSupport::isRenderableTextNode(const RenderObject* object)
{
    ASSERT(object->isText());
    // <br> is marked as text, but is not handled by the SVG layout code-path.
    const auto* svgInlineText = dynamicDowncast<RenderSVGInlineText>(object);
    return svgInlineText && !svgInlineText->hasEmptyText();
}
```

File: LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png
Comment (Ahmad-S792) [reply LayoutTests/platform/gtk/svg/text/textPathBoundsBug-expected.png:1]: Fix this in next push but might need to bit rebaselines.

**Comment by Ahmad-S792** [comment]
@nikolaszimmermann - Any update on above queries?

File: Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:363]: It is more of drive-by fix done by Blink developer, which is to remove duplicate check (one this and one below in 'while' case).

File: Source/WebCore/rendering/svg/SVGRenderSupport.h (hunk 2 above)
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/SVGRenderSupport.h:50]: Only `RenderSVGInline` and `RenderSVGText` use the new helper method. Both renderers share the same base class, `RenderLayerModelObject`, where we recently added SVG specific code with the intent to share it between SVG renderers that do not inherit from RenderSVGModelObject (which is the next best place to share code between related SVG renderers).

RenderSVGInline inherits from RenderInline : RenderBoxModelObject : RenderLayerModelObject : RenderElement : RenderObject, and RenderSVGText inherits from RenderSVGBlock : RenderBlockFlow : RenderBlock : RenderBox : RenderBoxModelObject : RenderLayerModelObject... Therefore RenderLayerModelObject is the nearest common ancestor.

I'd suggest to move the code there (there's already LBSE specific code there).

Comment (Ahmad-S792) [reply Source/WebCore/rendering/svg/SVGRenderSupport.h:50]: Just a note to myself for fixing that I can use 'lambda' like:
```
auto isEmptySVGInlineText = [](const RenderObject* object) {
        const auto svgInlineText =  dynamicDowncast<RenderSVGInlineText>(object);
        return svgInlineText && svgInlineText->hasEmptyText();
    };
```

**Pull request closed without merging**